data-processing/entities/sentences/extractor.py

```diff
@@ -123,12 +233,165 @@ def parse(self, tex_path: str, tex: str) -> Iterator[Sentence]:
+                    PATTERN_ABSTRACT_BEGIN, extended_tex_sub
+                )
+                tex_unit_dict["abstract_end"] = regex.findall(
+                    PATTERN_ABSTRACT_END, extended_tex_sub
+                )
+                tex_unit_dict["table_begin"] = regex.findall(
+                    PATTERN_TABLE_BEGIN, extended_tex_sub
+                )
+                tex_unit_dict["table_end"] = regex.findall(
+                    PATTERN_TABLE_END, extended_tex_sub
+                )
+                tex_unit_dict["figure_begin"] = regex.findall(
+                    PATTERN_FIGURE_BEGIN, extended_tex_sub
+                )
+                tex_unit_dict["figure_end"] = regex.findall(
+                    PATTERN_FIGURE_END, extended_tex_sub
+                )
+
+                tex_unit_dict["itemize_begin"] = regex.findall(
+                    PATTERN_ITEMIZE_BEGIN, extended_tex_sub
+                )
+                tex_unit_dict["itemize_end"] = regex.findall(
+                    PATTERN_ITEMIZE_END, extended_tex_sub
+                )
+
+                tex_unit_dict["ref"] = regex.findall(PATTERN_REF, extended_tex_sub)
+                tex_unit_dict["cite"] = regex.findall(PATTERN_CITE, extended_tex_sub)
+                tex_unit_dict["symbol"] = regex.findall(
+                    PATTERN_SYMBOL, extended_tex_sub, overlapped=False
+                )
+                tex_unit_dict["url"] = regex.findall(
+                    PATTERN_URL, extended_tex_sub, overlapped=False
+                )
+                tex_unit_dict["any"] = regex.findall(PATTERN_ANY, extended_tex_sub)
+
+                # store 'others' field by [any] - [section,label,...]
+                if len(tex_unit_dict["any"]) > 0:
+                    tex_unit_values_flattened = [
+                        fone
+                        for f in tex_unit_dict.keys()
+                        if f != "any"
+                        for fone in tex_unit_dict[f]
+                    ]
+                    others = []
+                    for any_field in tex_unit_dict["any"]:
+                        if any_field not in tex_unit_values_flattened:
+                            others.append(any_field)
+                    tex_unit_dict["others"] = others
+                tex_unit_dict.pop("any", None)
+
+                # store section information
+                if len(tex_unit_dict["abstract_begin"]) > 0:
+                    current_section = "ABSTRACT"
+                if len(tex_unit_dict["abstract_end"]) > 0:
+                    current_section = None
+                if len(tex_unit_dict["section"]) > 0:
+                    section = tex_unit_dict["section"][0]
+                    current_section = extract_text_from_tex_group(section)
+
+                # store figure/table information
+                # TODO @dykang add \label{} in table/figure for better matching
+                if len(tex_unit_dict["figure_begin"]) > 0:
```

Comment:
just noticing the repetitious code, it occurs to me that maybe a more succinct or clear rewriting would be something like:

```python
is_sentence_in_figure =tex_unit_dict["cite"] = (
  regex.findall(PATTERN_FIGURE_BEGIN, extended_tex_sub) and
  regex.findall(PATTERN_FIGURE_END, extended_tex_sub)
```

for each of the types of tex units. No biggie and not important, but just wanted to throw the idea out there.

Reply:
`regex.findall(PATTERN_FIGURE_BEGIN, extended_tex_sub)` only checks whether the current line includes \begin{figure} or not, since input tex only includes "single" line of tex, not a batch of entire tex lines for a figure. So, checking begin and end conditions together do not capture the block of lines in figure.